pilot/pkg/xds/sds.go

```diff
@@ -304,8 +309,58 @@ func toEnvoyCaSecret(name string, cert []byte) *discovery.Resource {
 	}
 }
 
-func toEnvoyKeyCertSecret(name string, key, cert []byte) *discovery.Resource {
-	res := util.MessageToAny(&envoytls.Secret{
+func toEnvoyKeyCertSecretWithPrivateKeyProvider(name string, key, cert []byte, pkpConf *mesh.PrivateKeyProvider) *anypb.Any {
+	var res *anypb.Any
+	switch pkpConf.GetProvider().(type) {
+	case *mesh.PrivateKeyProvider_Cryptomb:
+		crypto := pkpConf.GetCryptomb()
+		msg := util.MessageToAny(&cryptomb.CryptoMbPrivateKeyMethodConfig{
+			PollDelay: durationpb.New(time.Duration(crypto.GetPollDelay().Nanos)),
+			PrivateKey: &core.DataSource{
+				Specifier: &core.DataSource_InlineBytes{
+					InlineBytes: key,
+				},
+			},
+		})
+		res = util.MessageToAny(&envoytls.Secret{
+			Name: name,
+			Type: &envoytls.Secret_TlsCertificate{
+				TlsCertificate: &envoytls.TlsCertificate{
+					CertificateChain: &core.DataSource{
+						Specifier: &core.DataSource_InlineBytes{
+							InlineBytes: cert,
+						},
+					},
+					PrivateKeyProvider: &envoytls.PrivateKeyProvider{
+						ProviderName: "cryptomb",
+						ConfigType: &envoytls.PrivateKeyProvider_TypedConfig{
+							TypedConfig: msg,
+						},
+					},
+				},
+			},
+		})
+	default:
+		log.Warnf("Unknown PrivateKeyProvider")
+		return nil
+	}
+
+	return res
+}
+
+func toEnvoyKeyCertSecret(name string, key, cert []byte, pkpConf *mesh.PrivateKeyProvider) *discovery.Resource {
+	var res *anypb.Any
+	if pkpConf != nil {
+		res = toEnvoyKeyCertSecretWithPrivateKeyProvider(name, key, cert, pkpConf)
```

Comment:
style nit: I don't think we need these split code paths for `pkpConf != nil`. We can just have a big `switch` that has cases for `mesh.PrivateKeyProvider_Cryptomb`, <any future providers>, and `default`; the default case is the current code.

Essentially the default case *is* a PrivateKeyProvider